### PR TITLE
fix #1241

### LINKED ISSUE #1241: Check for negative values in BCECriterion
I have noticed that in [THNN/BCECriterion.c](https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/BCECriterion.c#L21) there is no assert for negative values, resulting in possible NaN loss and gradients.
I think it will be easier for developers to see their program crash (or catch the exception) with a message indicating that their values are negative instead of searching for the problem of their NaN loss.

### PR DESCRIPTION
**Pull request by stegben**

### CODE REVIEW

**Comment by apaszke** [comment]
Can you also add the checks to THCUNN?

File: torch/lib/THNN/generic/BCECriterion.c [hunk 1]
```diff
@@ -18,12 +18,16 @@ void THNN_(BCECriterion_updateOutput)(THNNState *state, THTensor *input,
       real x = *input_data;
       real y = *target_data;
       real w = *weights_data;
+      THNN_ARGCHECK(x < 0. || x > 1., 2, x,
```
Comment (apaszke) [inline torch/lib/THNN/generic/BCECriterion.c]: I think you wanted to use `THAssertMsg` here. The contbuilds are failing now.

**Comment by stegben** [comment]
Not familiar with CUDA computing, not sure if this code make sense.

**Comment by soumith** [comment]
there are two bugs in this PR. I've fixed them, and then merged this PR into master.
Thanks for the assertions.

**Comment by stegben** [comment]
THX!

**Pull request closed without merging**